### DISCUSSION TITLE
Understanding the Resource Usage

### DISCUSSION THREAD

**Discussion by HenningCode** [fastmachinelearning/hls4ml#533] [General]
Hello,

I want to understand the Resource usage of my Models:

I got this Model and I am creating the hls project with two different Reuse Factors. One time with the Maximum (12544,2048,1280) and one Time with a ReuseFactor of (14,16,16) 
```
    x = x_in = Input((28,28,1), name='input_1_m')
    x = Flatten(name='flatten_1_m')(x)
    x = QDense(16,
               kernel_quantizer=quantized_bits(8, 2, alpha=1),
               bias_quantizer=quantized_bits(8, 2, alpha=1),
               name='qdense_1_m')(x)
    x = QActivation("quantized_relu(8,4)", name='qact_1_m')(x)
    x = QDense(128,
               kernel_quantizer=quantized_bits(8, 2, alpha=1),
               bias_quantizer=quantized_bits(8, 2, alpha=1),
               name='qdense_2_m')(x)
    x = QActivation("quantized_relu(8,5)", name='qact_2_m')(x)
    x = QDense(10,
               kernel_quantizer=quantized_bits(8, 2, alpha=1),
               bias_quantizer=quantized_bits(8, 2, alpha=1),
               name='qdense_3_m')(x)
    x = Activation("softmax", name='softmax')(x)

    model = Model(inputs=[x_in], outputs=[x])
```
 These are the corresponding  resource usages: 
```
================================================================
== Utilization Estimates
================================================================
* Summary: 
+-----------------+---------+-------+--------+--------+-----+
|       Name      | BRAM_18K| DSP48E|   FF   |   LUT  | URAM|
+-----------------+---------+-------+--------+--------+-----+
|DSP              |        -|      -|       -|       -|    -|
|Expression       |        -|      -|       0|       2|    -|
|FIFO             |        1|      -|     119|     417|    -|
|Instance         |      248|    936|   49950|  151098|    -|
|Memory           |        -|      -|       -|       -|    -|
|Multiplexer      |        -|      -|       -|       -|    -|
|Register         |        -|      -|       -|       -|    -|
+-----------------+---------+-------+--------+--------+-----+
|Total            |      249|    936|   50069|  151517|    0|
+-----------------+---------+-------+--------+--------+-----+
|Available        |      624|   1728|  460800|  230400|   96|
+-----------------+---------+-------+--------+--------+-----+
|Utilization (%)  |       39|     54|      10|      65|    0|
+-----------------+---------+-------+--------+--------+-----+

================================================================
== Utilization Estimates
================================================================
* Summary: 
+-----------------+---------+-------+--------+--------+-----+
|       Name      | BRAM_18K| DSP48E|   FF   |   LUT  | URAM|
+-----------------+---------+-------+--------+--------+-----+
|DSP              |        -|      -|       -|       -|    -|
|Expression       |        -|      -|       0|       2|    -|
|FIFO             |        1|      -|     119|     417|    -|
|Instance         |       15|      3|   51228|   45978|    -|
|Memory           |        -|      -|       -|       -|    -|
|Multiplexer      |        -|      -|       -|       -|    -|
|Register         |        -|      -|       -|       -|    -|
+-----------------+---------+-------+--------+--------+-----+
|Total            |       16|      3|   51347|   46397|    0|
+-----------------+---------+-------+--------+--------+-----+
|Available        |      624|   1728|  460800|  230400|   96|
+-----------------+---------+-------+--------+--------+-----+
|Utilization (%)  |        2|   ~0  |      11|      20|    0|
+-----------------+---------+-------+--------+--------+-----+

```

Now for the Model with the Maximum Reuse for every Layer the DSPs are 3 which makes sense because for each layer one DSP is used. For the Model with Reuse of (14,16,16) the DSP usage is 936 which are only the DSP of the first layer 12544/14 = 896 and some more. Where are the DSPs for the other Layers? I am Calculating I should be using  12544/14 + 2048/16 + 1280/16 = 1104 DSPs. Furthermore I would like to understand why BRAM usages goes down that much. I though the weights are saved to the BRAM is that the case? 

Thanks for your help. 

I am using this branch if that helps: https://github.com/HenningCode/hls4ml/tree/zcu104

**Comment by vloncar** [comment]
You can read the per-layer reports in generated project directory, or open the project with Vivado HLS GUI and browse from there. For less than 9-10 bits, Vivado likes to use LUTs instead of DSPs, so you can check if that was the case.

As for the BRAMs, for a given reuse factor, we need to access weights efficiently, and BRAMs have only 1 or 2 ports, so the lower the RF, the more BRAMs we need to split the weights in. So it won't be n_weights * bits / 18k, except perhaps for the largest RF in some cases. Again, I recommend browsing the reports from Vivado HLS GUI to see how BRAMs were allocated.

**Reply by HenningCode** [reply]
Thank you very much! I will do that tomorrow and will come back if I got anymore questions. 

And I really like your project! I am writing a Master-Thesis right now about hls4ml, Vitis AI and embedded GPUs and hls4ml seems to be by far the best Option for smaller networks at least!

Edit: If I understand it correctly, thats also why the of  amount of LUTs used goes up? Because Vivado HLS replaces some DSPs with LUTs? Thats pretty interesting.
